game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/CheckGeneralBattleEnd.java

```diff
@@ -0,0 +1,130 @@
+package games.strategy.triplea.delegate.battle.steps.change;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.DiceRoll;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import games.strategy.triplea.delegate.battle.steps.RetreatChecks;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class CheckGeneralBattleEnd implements BattleStep {
+  private static final long serialVersionUID = 5172121497955756220L;
+
+  private final BattleState battleState;
+
+  private final BattleActions battleActions;
+
+  private final BiConsumer<IDelegateBridge, BattleState.Side> removeUnprotectedUnits;
+
+  private final Consumer<IDelegateBridge> retreatUnits;
+
+  @Override
+  public List<String> getNames() {
+    return List.of();
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.GENERAL_BATTLE_END_CHECK;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    if (hasSideLost(BattleState.Side.OFFENSE)) {
+      battleActions.endBattle(bridge);
+      battleActions.defenderWins(bridge);
+
+    } else if (hasSideLost(BattleState.Side.DEFENSE)) {
+      removeUnprotectedUnits.accept(bridge, BattleState.Side.DEFENSE);
```

Comment:
This step is, currently, the only step that needs access to logic from another battle step.  I'm undecided in how best to access that logic.  I've three thoughts:

1. Instantiate the steps and call the public method right when it is needed (aka `new RemoveUnprotectedUnits(battleState, battleActions).removeUnprotectedUnits(bridge, BattleState.Side.DEFENSE)`)
2. Pass in the the step to the constructor of `CheckGeneralBattleEnd` and then call the public method when it is needed.
3. Pass in a lambda of the public method and call the lambda.

I've currently coded up #3.

I don't like #2 because it will cause the other battle steps to be saved with this battle step and would make future changes harder to do.

I've tried #1 and it makes testing a little harder.